bindings/rust/src/lazy_reader.rs

```diff
@@ -0,0 +1,488 @@
+}
+
+pub const MAX_CACHE_SIZE: usize = 2048;
+pub const MIN_CACHE_SIZE: usize = 64;
+pub const NUM_T_SIZE: usize = 4;
+
+pub struct DataSource {
+    reader: Box<dyn Read>,
+
+    total_size: usize,
+    cache_start_point: usize,
+    // cache size may be smaller than cache.len()
+    cache_size: usize,
+    cache: Vec<u8>,
+}
+
+#[derive(Clone)]
+pub struct Cursor {
+    pub offset: usize,
+    pub size: usize,
+    pub data_source: Rc<RefCell<DataSource>>,
+}
+
+pub struct Union {
+    pub item_id: usize,
+    pub cursor: Cursor,
+}
+
+pub fn read_at(cur: &Cursor, buf: &mut [u8]) -> Result<usize, Error> {
+    let read_len = min(cur.size, buf.len());
+    let ds = &mut *cur.data_source.borrow_mut();
+    if read_len > ds.cache.len() {
+        return ds.reader.read(buf, cur.offset);
+    }
+    if cur.offset < ds.cache_start_point
+        || (cur.offset + read_len) > (ds.cache_start_point + ds.cache_size)
+    {
+        let reader = &ds.reader;
+        let size = reader.read(&mut ds.cache[..], cur.offset).unwrap();
+        if size < read_len {
+            return Err(Error::Read(format!(
+                "read_at `if size({}) < read_len({})`",
+                size, read_len
+            )));
+        }
+        ds.cache_size = size;
+        ds.cache_start_point = cur.offset;
+
+        if ds.cache_size > ds.cache.len() {
+            return Err(Error::Read(format!(
+                "read_at `if ds.cache_size({}) > ds.cache.len()({})`",
+                ds.cache_size,
+                ds.cache.len()
+            )));
+        }
+    }
+    if cur.offset < ds.cache_start_point || (cur.offset - ds.cache_start_point) > ds.cache.len() {
+        return Err(Error::Read(
+            "read_at `if cur.offset < ds.start_point || ...`".into(),
+        ));
+    }
+    let read_point = cur.offset - ds.cache_start_point;
+    if read_point + read_len > ds.cache_size {
+        return Err(Error::Read(
+            "read_at `if read_point + read_len > ds.cache_size`".into(),
+        ));
+    }
+    buf.copy_from_slice(&ds.cache[read_point..(read_point + read_len)]);
+    Ok(read_len)
+}
+
+impl Cursor {
+    /**
+    cache_size: normally it can be set to MAX_CACHE_SIZE(2K)
+    total_size: the size of cursor. If it's set a smaller value,
+    `out of bound` will occur when `reader` try to read the data beyond that.
+    reader: interface to read underlying data
+     */
+    pub fn new(total_size: usize, reader: Box<dyn Read>) -> Self {
+        let data_source = DataSource {
+            reader,
+            total_size,
+            cache_start_point: 0,
+            cache_size: 0, // when created, cache is not filled
+            cache: vec![0u8; MAX_CACHE_SIZE],
+        };
+        Cursor {
+            offset: 0,
+            size: total_size,
+            data_source: Rc::new(RefCell::new(data_source)),
+        }
+    }
+
+    pub fn add_offset(&mut self, offset: usize) {
+        self.offset = self.offset.checked_add(offset).unwrap();
```

Comment:
should returning error instead of unwrap, otherwise, the user could pass an illegal index that would cause an unwrap error:
```
GeneratedReader.get(usize::MAX)
  => self.cursor.slice_by_offset
  => unwrap error
```

Reply:
I think it's ok to unwrap directly. When error occurs, it means the result of operator "+" is overflow. When we are writing code like below:
```Rust
let c = a + b;
```
It's not needed to return an Error.

Reply:
They're not the same errors, your example will throw overflow error, but in this PR, unwrap on the checked_xxx op will throw `None unwrap` error, it's difficult to debug.